src/stream/src/error.rs

```diff
@@ -20,13 +20,20 @@ use risingwave_pb::PbFieldNotFound;
 use risingwave_rpc_client::error::ToTonicStatus;
 use risingwave_storage::error::StorageError;
 
-use crate::executor::StreamExecutorError;
+use crate::executor::{Barrier, StreamExecutorError};
+use crate::task::ActorId;
 
 /// A specialized Result type for streaming tasks.
 pub type StreamResult<T> = std::result::Result<T, StreamError>;
 
 /// The error type for streaming tasks.
-#[derive(thiserror::Error, Debug, thiserror_ext::Box)]
+#[derive(
+    thiserror::Error,
+    Debug,
+    thiserror_ext::Arc,
+    thiserror_ext::ContextInto,
+    thiserror_ext::Construct,
+)]
```

Comment:
Why `Arc`

What is `ContextInto`

Reply:
> Why `Arc`

The error is cloned here: https://github.com/risingwavelabs/risingwave/blob/f5efe3f3a7a7d1271cf60a1eb4f56e64c18ad569/src/stream/src/task/barrier_manager/managed_state.rs#L179-L181

> What is `ContextInto`

It's another kind of constructor derivation! 😄 Basically, we can write
https://github.com/risingwavelabs/risingwave/blob/f5efe3f3a7a7d1271cf60a1eb4f56e64c18ad569/src/stream/src/task/barrier_manager/managed_state.rs#L167-L168

as a sugar of 
```rust
let err = StreamError::unexpected_exit(err, actor_id);
```

It's not that sweet here, but can be useful for transforming the `Result`s as it's also implemented on `Result`s.

https://github.com/risingwavelabs/risingwave/blob/f5efe3f3a7a7d1271cf60a1eb4f56e64c18ad569/src/common/heap_profiling/src/jeprof.rs#L49-L52

It's highly inspired by `anyhow::Context` and [`snafu::Context`](https://docs.rs/snafu/latest/snafu/trait.ResultExt.html#tymethod.context). I'll write the docs for them ASAP.

Reply:
Oh, the context here is always evaluated. 😕 Perhaps I need to introduce a `with` method.

https://github.com/risingwavelabs/risingwave/blob/f5efe3f3a7a7d1271cf60a1eb4f56e64c18ad569/src/common/heap_profiling/src/jeprof.rs#L49-L52